Stop retrying after a number of attempt

cyberdelia opened the pull request:
Stop retrying after a number of attempt, sometimes the problem will not fix itself after a certain number of retries but restarting ``wal-e`` will (or else).

This is unfinished, as we should probably still allow to retry indefinitely and allow to change the number of attempts, so take this as a conversation starter.

fdr commented on wal_e/retries.py:
```diff
@@ -70,6 +71,8 @@ def shim(*args, **kwargs):
                         exc_processor_cxt = exception_processor(
                             exception_info_tuple,
                             exc_processor_cxt=exc_processor_cxt)
+                    except UserCritical:
```
This will unfortunately change the semantics of any `UserCritical`. Have you checked all such `raise`s to make sure that they make sense to terminate a retry-loop?

cyberdelia replied:
I mostly have used ``UserCrticial`` as a shortcut. 
Note that only ``UserCritical`` exceptions raised by an ``exception_processor`` will be taken into account, and all of them are just logging errors.

cyberdelia commented:
I introduced exponential backoff (capped at 2 minutes), and moved the logic inside the ``retry`` decorator.

Should ``max_retries`` be only env and not also an option?

fdr commented on wal_e/retries.py:
```diff
@@ -74,6 +86,11 @@ def shim(*args, **kwargs):
                         del exception_info_tuple
+
+                    # Exponential backoff with jitter capped at 2 minutes.
+                    duration = min(120, (1 * 2 ** retries)) / 2
```
why 1 * 2?

cyberdelia replied:
I really like useless math operations.
